core/trino-spi/src/main/java/io/trino/spi/predicate/Utils.java

```diff
@@ -64,4 +72,53 @@ static RuntimeException handleThrowable(Throwable throwable)
         }
         return new RuntimeException(throwable);
     }
+
+    /**
+     * Try to expand {@code valueSet} into a discrete set of (at most {@code limit}) objects.
+     * For example: [1, 5] can be expanded into {1, 2, 3, 4, 5}.
+     * If the data type is not supported or the expansion results in too many values, {@code Optional.empty()} is returned.
+     */
+    public static Optional<Collection<Object>> tryExpandRanges(ValueSet valuesSet, int valuesLimit)
+    {
+        if (!(valuesSet instanceof SortedRangeSet) || valuesLimit <= 0) {
+            return Optional.empty();
+        }
+        List<Range> ranges = valuesSet.getRanges().getOrderedRanges();
+        Type type = valuesSet.getType();
+        List<Object> result = new ArrayList<>(valuesLimit);
+        for (Range range : ranges) {
+            if (range.isSingleValue()) {
+                result.add(range.getSingleValue());
+                if (result.size() > valuesLimit) {
+                    return Optional.empty();
+                }
+                continue;
+            }
+            if (range.isLowUnbounded() || range.isHighUnbounded()) {
+                return Optional.empty();
+            }
+            Optional<Stream<?>> discreteValues = type.getDiscreteValues(new Type.Range(range.getLowBoundedValue(), range.getHighBoundedValue()));
+            if (discreteValues.isEmpty()) {
+                return Optional.empty();
+            }
+            Iterator<?> iterator = discreteValues.get().iterator();
+            if (!range.isLowInclusive()) {
+                if (!iterator.hasNext()) {
+                    throw new IllegalStateException("discreteValues iterator is empty");
+                }
+                iterator.next();
+            }
+            int valuesToDrop = range.isHighInclusive() ? 0 : 1;
+            while (iterator.hasNext()) {
```

Comment:
I think this whole loop can be simplified to:

```java
while (iterator.hasNext()) {
    Object current = iterator.next();

    if (range.isHighInclusive() || iterator.hasNext()) {
        result.add(current);
    }

    if (result.size() > valuesLimit) {
        return Optional.empty();
    }
}
```

Reply:
Looks good, updated the code :)